src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -241,77 +251,99 @@ public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext conte
 
-                        variantDefinition = await GetAssignedVariantAsync(
-                            featureDefinition,
-                            targetingContext,
-                            cancellationToken)
-                            .ConfigureAwait(false);
+                        if (targetingContext != null)
+                        {
+                            variantDefinition = await AssignVariantAsync(evaluationEvent, targetingContext, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (variantDefinition == null)
+                        {
+                            variantDefinition = evaluationEvent.FeatureDefinition
+                                .Variants
+                                .FirstOrDefault(variant =>
+                                    variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenEnabled);
+
+                            evaluationEvent.VariantAssignmentReason = VariantAssignmentReason.DefaultWhenEnabled;
+                        }
                     }
+                    else
+                    {
+                        variantDefinition = evaluationEvent.FeatureDefinition
+                            .Variants
+                            .FirstOrDefault(variant =>
+                                variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenDisabled);
+
+                        evaluationEvent.VariantAssignmentReason = VariantAssignmentReason.DefaultWhenDisabled;
+                    }
+
+                    evaluationEvent.Variant = variantDefinition != null ? GetVariantFromVariantDefinition(variantDefinition) : null;
 
-                    if (variantDefinition != null && featureDefinition.Status != FeatureStatus.Disabled)
+                    //
+                    // Override IsEnabled if variant has an override
+                    if (variantDefinition != null && evaluationEvent.FeatureDefinition.Status != FeatureStatus.Disabled)
                     {
                         if (variantDefinition.StatusOverride == StatusOverride.Enabled)
                         {
-                            isFeatureEnabled = true;
+                            evaluationEvent.Enabled = true;
                         }
                         else if (variantDefinition.StatusOverride == StatusOverride.Disabled)
                         {
-                            isFeatureEnabled = false;
+                            evaluationEvent.Enabled = false;
                         }
                     }
                 }
-            }
 
-            foreach (ISessionManager sessionManager in _sessionManagers)
-            {
-                await sessionManager.SetAsync(feature, isFeatureEnabled).ConfigureAwait(false);
-            }
+                Debug.Assert(evaluationEvent.Variant != null ? evaluationEvent.VariantAssignmentReason != VariantAssignmentReason.None : true);
```

Comment:
I've never seen an assert like this before. Seems a bit odd.

Reply:
Removed.